### ISSUE TITLE
Regression: New formatter block formatting rules leaves ending tags of nested HTML block elements on same line; breaks existing formatting functionality.

### ISSUE THREAD

**Issue by garretwilson** [eclipse-jdt/eclipse.jdt.core#1131]
This is caused by the changes made for #991.

I upgraded to Eclipse EE 2023-06 RC1 on Windows 10. With the new "Do not put block tags on separate lines" enabled, the formatter  produces this for HTML in Javadocs (`/*` and `*` and the like removed for clarity, to concentrate on the HTML structure):

```html
<p>For example for the file <code>path/to/example.foo.bar</code> the following would be returned in order:</p>
<ol>
<li><code>foo.bar</code></li>
<li><code>bar</code></li></ol>
```

Note that the ending tag `</ol>` should be on a new line, like this:

```html
<p>For example for the file <code>path/to/example.foo.bar</code> the following would be returned in order:</p>
<ol>
<li><code>foo.bar</code></li>
<li><code>bar</code></li>
</ol>
```

I've discussed the problems with the current algorithm and the what the algorithm should be in #991 and in [Bug 471920](https://bugs.eclipse.org/bugs/show_bug.cgi?id=471920) before that. But let me just make a comment on the [configuration mentioned in another comment](https://github.com/eclipse-jdt/eclipse.jdt.core/issues/991#issuecomment-1520858735):

```java
	String formatCodeTags = "(pre)"; //$NON-NLS-1$
	String separateLineTags = "(nl|table|tr)"; //$NON-NLS-1$
	String breakBeforeTags = "(dd|dt|li|td|th|h1|h2|h3|h4|h5|h6|q)"; //$NON-NLS-1$
	String blockTags = "(p|dl|ul|ol|hr|dir)"; //$NON-NLS-1$
	String breakAfterTags = "(br)"; //$NON-NLS-1$
	String noFormatTags = "(code|tt)"; //$NON-NLS-1$
	String otherTags = "([\\S&&[^<>]]++)"; //$NON-NLS-1$
```

I realize that this code probably has a long history with changes by many people, so I'm not criticizing the latest changes that were made to try to improve it. But conceptually this is super way too complicated, in my opinion. If the code were being written from scratch, one wouldn't need separate `breakBeforeTags` and `blockTags` as far as line-breaks go for HTML. (For example, why is an `h1` being treated differently from a `p` as far as line breaks go?) Simplified only slightly, one would just need **a single `blockElements`**. The algorithm would look like this:

1. Is this the beginning tag of an element in `blockElements`? If so, put a newline _before_ it.
2. Is this the ending tag of of an element in `blockElements`? If so, put a newline _after_ it.

And that is basically it.

**Comment by garretwilson** [comment]
It looks like this is still broken in Eclipse 2023-09 RC1. @jjohnstn can you give me a status on this? I was so excited after your work on #991, thinking that after so many years it might finally be addressed. But I'm not sure if any progress has been made since Eclipse 2023-06. Any news?

**Comment by garretwilson** [comment]
Oh no! It would appear that the changes made in #991 actually caused a regression in the _existing_ formatting!

Before we would have Javadocs like this:

```java
/**
 * <p>
 * blah blah blah
 * </p>
 * 
 * <pre>
 * <code>com.example.FooConcernProvider</code>
 * </pre>
 * <p>
 * more blah blah blah
 * </p>
 */
```

Now after #991, at least as of Eclipse 2023-09 RC1, Eclipse will _unwrap_ the `</pre>` block tag, resulting in this:

```java
/**
…
 * <pre>
 * <code>com.example.FooConcernProvider</code> </pre>
…
 */
```

See Javadocs in [Csar.java](https://github.com/globalmentor/csar/blob/main/src/main/java/io/csar/Csar.java) as an example.

Now after almost a decade, not only is the issue not fixed, it is now broken worse than it was before. 😢

If you can't fix it, can at least reverse the changes in #991 so that _existing_ formatting won't be broken?

**Comment by garretwilson** [comment]
> I'll take a look at it once all the postponed 4.30 M1 changes (too late for M3) are merged.

Fair enough. Thank you!
